### ISSUE TITLE
[no-unused-vars] false positive no-unused-vars for lambda type's parameter name or interface function parameter name

### ISSUE THREAD

**Issue by anatawa12** [typescript-eslint/typescript-eslint#3472]
- [x] I have tried restarting my IDE and the issue.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

I'm using 4.26.0

**Repro**

``

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": "error"
  }
}
```

```TS
// your repro code case

interface SomeInterface {
  // the no-unused-vars says `ESLint: 'dir' is defined but never used.(@typescript-eslint/no-unused-vars)`
  serverProvider(dir: string): Promise<void>
}

// the no-unused-vars says `ESLint: 'dir' is defined but never used.(@typescript-eslint/no-unused-vars)`
type SomeLambdaType = (dir: string) => Promise<void>
```

**Expected Result**

No error will happen because they can't be used / they generally not used.

**Actual Result**

no-unused-vars error happans.

**Additional Info**

**Versions**

Accoring to `package-lock.json` (version 2) or `node -v`

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.26.0` |
| `@typescript-eslint/parser`        | `4.26.0` |
| `TypeScript`                       | `4.3.2` |
| `ESLint`                           | `7.27.0` |
| `node`                             | `v12.19.0` |

**Comment by bradzacher** [comment]
We have passing tests for this in master.

https://github.com/typescript-eslint/typescript-eslint/blob/b814e635c1f34139c89e0176727480935dd45cac/packages/eslint-plugin/tests/rules/no-unused-vars/no-unused-vars.test.ts#L658-L669

Likely you have an old version installed.

**Comment by timweightman** [comment]
@bradzacher Hate to be a pain here- any chance you can shed a little extra light on this?

I've got same package versions as @anatawa12 and getting the same error on node v15.14.0

Which version of which package is considered "old", I installed all of these on a project within the past week (except node).

package | version | published
-- | -- | --
@typescript-eslint/eslint-plugin | 4.26.0 | [1 June 2021](https://www.npmjs.com/package/@typescript-eslint/eslint-plugin)
@typescript-eslint/parser | 4.26.0 | [1 June 2021](https://www.npmjs.com/package/@typescript-eslint/parser)
TypeScript | 4.3.2 | [27 May 2021](https://www.npmjs.com/package/typescript)
ESLint | 7.27.0 | [22 May 2021](https://www.npmjs.com/package/eslint)
node | v15.14.0 | [06 April 2021](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V15.md#15.14.0)

**Comment by bradzacher** [comment]
Not really - no.
The versions don't provide me much information into your setup.

The best way for me to help you is for you to create an isolated repro repo.
Then we can talk apples to apples as I can exactly see your problem.

**Comment by vith** [comment]
I was also hitting this problem due to the latest `react-scripts@4.0.3` depending on outdated versions of `@typescript-eslint` packages (`^4.5.0`), despite my project having explicit dependencies on the latest versions (`^4.26.1`).

I've had to add this to my `package.json` to work around it (using `yarn`):

```json
  "resolutions": {
    "@typescript-eslint/eslint-plugin": "^4.26.1",
    "@typescript-eslint/parser": "^4.26.1"
  },
```

Seems OK now.
